### ISSUE TITLE
Specification using enum values does not follow `@Enumerated(EnumType.STRING)` and tries to convert it to smallint

### ISSUE THREAD

**Issue by jluiz20** [spring-projects/spring-data-jpa#3486]
I have been migrating some services from Spring Boot 2.7.18 to Spring Boot 3.2.5 and after it, I am facing a specific issue with an enum column only using the Specification and I think it could be a bug (or I am doing something wrong).

I have an entity with the following field

```java
    @Enumerated(EnumType.STRING)
    @Builder.Default
    @Column(nullable = false)
    var acceptanceStatus: PaymentAcceptanceStatus = PaymentAcceptanceStatus.RECEIVED,
```
As the annotation shows, I am storing it as a String in the DB (PostgreSQL) and that works perfectly. 

![image](https://github.com/spring-projects/spring-data-jpa/assets/9250618/7218c291-b5bd-4061-b0ac-df4c0ebeb0a9)

It also works well when fetching the entity by id using normal repository methods.

But when I create a Specification filtering by the enum, like below.

![image](https://github.com/spring-projects/spring-data-jpa/assets/9250618/158e583a-177e-4c6c-9d64-3de319126048)

I get an error `ERROR: invalid input syntax for type smallint: "SENT"`

In debugging further, it seems that it is trying to convert the value passed `SENT` to a smallint. This is the query generated (some other fields were truncated)

`org.hibernate.exception.DataException: JDBC exception executing SQL [select prw1_0.id,prw1_0.acceptance_status,prw1_0.client_id from my_table prw1_0 where cast(prw1_0.acceptance_status as smallint)=? and cast(prw1_0.acceptance_status as smallint)=? order by prw1_0.received_at desc offset ? rows fetch first ? rows only] [ERROR: invalid input syntax for type smallint: "SENT"] [n/a] `

going deeper it seems that in the `SimpleJpaRepository` when building the parameterBindingMap, it is defining the `jdbcType` as `TyneIntJdbcType` instead of some kind of varchar or string that I would expect based on the `@Enumerated(EnumType.STRING)` definition.

![image](https://github.com/spring-projects/spring-data-jpa/assets/9250618/5de16e9a-3fed-4409-ba22-879b4b05d893)

Now, I am not sure if I need to add some other annotation of metadata to the field, or if the problem is the Hibernate or in the data jpa. I came here as saving and fetching are working so I think the mapping of the field is correct in Hibernate and I only have this problem when trying to use the Specification.

I also forced the latest version `3.3.0` of the library and I still see the problem. Also, tried to force a column definition     `@Column(nullable = false, columnDefinition = "varchar(255)")` but it didn't help.

Anything I can do? Any help is really appreciated! Thank you!

**Comment by quaff** [comment]
Could you provider a minimal reproducer project?

**Comment by jluiz20** [comment]
Thank you for your attention and I really sorry as I found out that the problem is on my side.

Turns out that we have a custom code that builds Specifications based on generic parameters and due to the Hibernate change it started breaking.

The previous code was creating an `Expression` using 
```java
 pathExpression = path.as(path.getJavaType());
```

The implementation was `SingularAttributePath` and it was mapped correctly, 

On hibernate 6, the same logic was creating a `SelfRenderingSqmFunction` with the incorrect types. 

![image](https://github.com/spring-projects/spring-data-jpa/assets/9250618/e3d214b8-2b11-4656-87cf-5807ca336a36)

Turns out that I only need the `.as(path.getJavaType())` anymore and the new logic is
```java
 pathExpression = path;
```
which maps the correct type

![image](https://github.com/spring-projects/spring-data-jpa/assets/9250618/8ed34e52-40e2-454d-a461-2b599b11d762)

I honestly don't know how this magic works, but it is not working again.

Sorry for creating an issue that was a problem on my side.
